main/src/main/scala/sbt/internal/LayeredClassLoaders.scala

```diff
@@ -10,119 +10,253 @@ package sbt.internal
 import java.io.File
 import java.net.URLClassLoader
 import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.{ AtomicBoolean, AtomicReference }
 
 import sbt.internal.inc.classpath._
 import sbt.io.IO
 
 import scala.collection.JavaConverters._
-import scala.collection.mutable.ListBuffer
 
-private[sbt] class LayeredClassLoaderImpl(
+/**
+ * A simple ClassLoader that copies native libraries to a temporary directory before loading them.
+ * Otherwise the same as a normal URLClassLoader.
+ * @param classpath the classpath of the url
+ * @param parent the parent loader
+ * @param tempDir the directory into which native libraries are copied before loading
+ */
+private[internal] class LayeredClassLoaderImpl(
     classpath: Seq[File],
     parent: ClassLoader,
-    override protected val resources: Map[String, String],
-    tempDir: File,
-) extends URLClassLoader(classpath.toArray.map(_.toURI.toURL), parent)
-    with RawResources
-    with NativeCopyLoader
-    with AutoCloseable {
-  private[this] val nativeLibs = new java.util.HashSet[File]().asScala
-  override protected val config = new NativeCopyConfig(
-    tempDir,
-    classpath,
-    IO.parseClasspath(System.getProperty("java.library.path", ""))
-  )
-  override def findLibrary(name: String): String = {
-    super.findLibrary(name) match {
-      case null => null
-      case l =>
-        nativeLibs += new File(l)
-        l
+    tempDir: File
+) extends URLClassLoader(classpath.map(_.toURI.toURL).toArray, parent)
+    with NativeLoader {
+  setTempDir(tempDir)
+}
+
+/**
+ * This classloader doesn't load any classes. It is able to create a two layer bundled ClassLoader
+ * that is able to load the full project classpath. The top layer is able to directly load the
+ * project dependencies. The bottom layer can load the full classpath of the run or test task.
+ * If the top layer needs to load a class from the bottom layer via java reflection, we facilitate
+ * that with the `ReverseLookupClassLoader`.
+ *
+ *
+ * This holder caches the ReverseLookupClassLoader, which is the top loader in this hierarchy. The
+ * checkout method will get the RevereLookupClassLoader from the cache or make a new one if
+ * none is available. It will only cache at most one so if multiple concurrently tasks have the
+ * same dependency classpath, multiple instances of ReverseLookupClassLoader will be created for
+ * the classpath. If the ReverseLookupClassLoader makes a lookup in the BottomClassLoader, it
+ * invalidates itself and will not be cached when it is returned.
```

Comment:
I am a bit confused here.
Let say there's subproject A, B, and C with identical libraries using Java reflection, and you run all three concurrently:
- checkout A
- checkout B
- checkout C
- loadClass A
- loadClass B
- loadClass C
- checkin A
- checkin B
- checkin C

Doesn't the first loadClass set it dirty, but B won't know?

Reply:
The cache has size one because I didn’t think it was worth it to keep more than one instance hot. In the sequence described above, project A gets the cached loader first, then a new loader is created for each of B and C. t. Let’s say only project A dirties the loader. When it checks in its loader, it gets closed and discarded because it was marked dirty. Then when B checks in its loader, that becomes the cached instance. When C checks in its loader, it gets closed and discarded because the cache was already filled by B.

Reply:
Because sbt runs tests in parallel, this could probably happen with one subproject too.

Reply:
> The cache has size one because I didn’t think it was worth it to keep more than one instance hot.

ok. That makes sense.

Reply:
Yeah. I figured that in that case this is no worse than having ScalaLibrary. The main issue is that ClassLoaderCache as presently written doesn’t support having multiple loaders for the same classpath. That is essential for the dependency layer with java reflection or else in the example above, the loader for subproject a could load classes for subproject b. This can get really complicated if one tries to design the optimal solution. I think the current implementation is a reasonable compromise without adding additional strategies (for example there could be AllLibraryJars and AllLibraryJarsWithReflection). The other thing I thought about trying to sneak in was an exclude filter for the jars in the dependency layer. I am not sure what the right api is though so it felt more like a 1.4.0 feature.

Reply:
So B's singleton could be reused by A too, which could be tricky if the state is dependent on some external factor like application.conf.

Reply:
Oh, so in that case projects a and b would not share a dependency layer because the resource loader goes between scala library/reflect and the dependency layer. Projects a and b would both have to have empty resource paths for that loader to be shared between them so if application.conf is different, they would be completely isolated.

Reply:
I forgot about that but I added the resource layer specifically for this reason.